### DISCUSSION TITLE
Фильтр по регулярному выражению для исполнителя

### DISCUSSION THREAD

**Discussion by Bondian** [Chimildic/goofy#79] [Помощь с алгоритмом] [answered]
Фильтрую треки и альбомы содержащие кириллицу
`Filter.match(Tracks, '[а-яА-ЯёЁ]+', false);`
Фильтр работает только по названию трека и альбому.
Не могу разобраться, как добавить в результат  треки с исполнителями, прописанными кириллицей, у которых название трека и альбома прописаны латиницей. Это вообще возможно?

**Comment by amd64fox** [comment]
Чуть ниже [обсуждалась](https://github.com/Chimildic/goofy/discussions/55) похожая проблема.
Нужно создать функцию [по инструкции](https://github.com/Chimildic/goofy/discussions/18), чтобы не потерять при обновлении.
```js
Filter.matchArtist = function (items, strRegex, invert = false) {
    let regex = new RegExp(strRegex, 'i');
    let filteredTracks = items.filter((item) => {
        if (typeof item == 'undefined') {
            return false;
        } else if (item.hasOwnProperty('album')) {
            return (
                invert ^
                (regex.test(item.artists[0].name.formatName()))
            );
        }
        return invert ^ regex.test(item.name.formatName());
    });
    Combiner.replace(items, filteredTracks);
};
```
С помощью  `Filter.matchArtist` фильтруем  исполнителей с кириллицей, потом используем 2 проход через привычный вам `Filter.match` по треку и альбому.

**Reply by Bondian** [reply]
Спасибо. Пока пытаюсь разобраться с предложенным ниже методом... Позже внимательно изучу.

**Comment by Chimildic** [comment, marked as answer]
В последних версиях исполнитель учитывается без дополнительных правок.

**Старый ответ**
Можно объединить вместе (трек, альбом, исполнитель). Также [по инструкции](https://github.com/Chimildic/goofy/discussions/18) добавив следующую функцию
```js
// Строго: все части должны иметь кириллицу
Filter.matchAll = function (tracks, strRegex) {
    let regex = new RegExp(strRegex, 'i');
    let filteredTracks = tracks.filter((track) => {
        return regex.test(track.name.formatName()) && 
               regex.test(track.album.name.formatName()) &&
               regex.test(track.artists[0].name.formatName());
    });
    Combiner.replace(tracks, filteredTracks);
}

// Не строго: хотя бы где-то есть кириллица
Filter.matchAll = function (tracks, strRegex) {
    let regex = new RegExp(strRegex, 'i');
    let filteredTracks = tracks.filter((track) => {
        return regex.test(track.name.formatName()) || 
               regex.test(track.album.name.formatName()) ||
               regex.test(track.artists[0].name.formatName());
    });
    Combiner.replace(tracks, filteredTracks);
}
```


И вызвать подобным образом. 
```js
Filter.matchAll(tracks, '^[а-я0-9 ]+$');
```
Регулярное выражение будет отбирать только кириллицу (возможно с числами) и пробелы (для нескольких слов). Остальные спец символы удаляются при вызове `formatName` и не играют роли при проверке.

**Reply by Bondian** [reply]
Спасибо!
Выбрал функцию Filter.matchAll с комментарием: "Не строго: хотя бы где-то есть кириллица" (именно это и нужно, чтобы хоть одна кириллическая буква присутствовала хотя бы где-то). Сохранил по инструкции в новом файле прямо под library.
Использовал не строгое регулярное выражение, которое ранее правильно фильтровало по имени трека и альбома.
Протестировал с плейлистом. 
И похоже что-то делаю не так... 
Трек с исполнителем "масло черного тмина" с именем трека и альбомом на латинице не отфильтровался.
Приведу код полностью, чтобы была возможность разобраться

```
function KyrillicFilterTest() {
  let tracks = Source.getPlaylistTracks('Набор треков для теста', '74LtHgzZciefv6qETNoHOc');

     Filter.matchAll(tracks, '[а-яА-ЯёЁ]+', false); // подходит для "хотя бы одна латинская буква хоть где-то"
     // Filter.matchAll(tracks, '^[а-я0-9 ]+$'); // предложено для строгой фильтрации латиницы.

      Playlist.saveWithReplace({
       id: '2Uj25Ofadu11X4hG2YMC8M', // после первого создания плейлиста
       name: 'Результат теста',
       // public: false,
       tracks: tracks,
       description: Playlist.getDescription(tracks),
       });
}
```

**Reply by Chimildic** [reply]
>именно это и нужно, чтобы хоть одна кириллическая буква присутствовала хотя бы где-то

> Трек с исполнителем "масло черного тмина" с именем трека и альбомом на латинице не отфильтровался

Сами себе противоречите

**Reply by Bondian** [reply]
В силу отсутствия опыта в описании логики кода, я неудачно выразился . Тут сыграла шутку многозначность русского языка. "**_Отфильтрованная_** чистая вода" и "**_отфильтрованная_** примесь" - две логические противоположности.
Я имел  ввиду, что трек группы "масло черного тмина" не стал "чистой водой" результирующего плейлиста, куда попали треки, которые содержали хоть одну кириллическую букву в названии трека или альбома (такой же результат, как до создания новой функции). 
Не знаю, в каких терминах это можно коротко описать языком программиста, поэтому вышло так витиевато...
Делаю вывод - либо регулярное выражение до сих не "не видит" названия исполнителя, либо само регулярное выражение не подходит....

**Reply by Chimildic** [reply]
Как оказалось, трек от этой группы возвращается от Spotify с именем исполнителя `maslo chernogo tmina`. Поэтому не попадает в результат по регулярному выражению. Почему остальных так не преобразует - загадка. 

Вероятно нашел решение. К запросу треков можно добавить параметр локали, чтобы получать корректный язык. 
Сделаю это несколько позднее. Это задача для обновления кода библиотеки, а не функции `matchAll`.

**Reply by Bondian** [reply]
Превелико благодарю! Всё понятно!

**Reply by Chimildic** [reply]
Тестируем в версии 1.4.8

**Reply by efremych** [reply]
Правильно понимаю, что на запрос Любимых треков нововведение не распространяется?
И второй вопрос, Last Fm в каком виде получает треки (например, для истории)?

**Reply by Chimildic** [reply]
@efremych сейчас попробовал добавить локаль на все get-запросы без разбора, а не только плейлисты и история. На лишние аргументы Spotify вроде не ругается. 
Про ластфм не понял вопрос.

**Reply by efremych** [reply]
Про Ласт фм - попадают ли в кэш истории исполнители уже локализованными. Тщательно не смотрел, но вроде не смог найти в истории уже «локализированные» версии исполнителей…
 При том, что в кэше спотифая исполнители локализированы (и так понимаю, надо пройтись по файлу и переименовать записи до обновления библиотеки :) благо 2 исполнителя у меня таких нашлось). 
 
Я уже менял у себя запрос любимых треков (потом вернул обратно) и всё равно есть лишние записи, потому вот грешу пока на Историю. У меня просто в файле подсчитываются проигрывания треков, вот и увидел неудачные дедубликации - локализированные и нет версии треков…

**Reply by Chimildic** [reply]
С [коммитом](https://github.com/Chimildic/goofy/commit/1337850365fd20863031b0368586bd3386cf22ba), о котором сказал выше - везде теперь должна быть ру-локаль.

В удалении дубликатов был баг, не удалял по одинаковому id. Недавно [исправлено](https://github.com/Chimildic/goofy/commit/b0a203b1b71bb080dbb7ee01b298b80ed6c28937). Поэтому руками исправлять кэш не нужно.

**Reply by efremych** [reply]
>  везде теперь должна быть ру-локаль.

Да, теперь и в кэш ластфм пишется корректно.
На всякий случай переименовал старые записи кэшей истории :) История увеличилась на 3 трека))) Странное число, ну, да, ладно) Лишние строчки из топа проигрываний исчезли. Теперь точно всё хорошо должно быть.
Спасибо!
